File: mobile/lib/domain/repository/category_repository.dart
import '../model/category.dart';
import '../model/enums.dart';

abstract class CategoryRepository {
  Future<void> create(Category category);
  Future<Category?> findById(String id);
  Future<List<Category>> listByFamily(
    String familyId, {
    CategoryType? type,
    bool includeArchived = false,
  });
  Future<void> update(Category category);
  Future<void> softDelete(String id);
  Stream<List<Category>> watchByFamily(String familyId);
}

File: backend/src/main/kotlin/mymoney/domain/repository/CategoryRepository.kt
package mymoney.domain.repository

import mymoney.domain.model.Category
import mymoney.domain.model.CategoryType
import java.util.UUID

interface CategoryRepository {
    suspend fun create(category: Category): Category
    suspend fun findById(id: UUID): Category?
    suspend fun listByFamily(
        familyId: UUID,
        type: CategoryType? = null,
        includeArchived: Boolean = false,
    ): List<Category>
    suspend fun update(category: Category): Category

    /**
     * Soft-deletes a category. Callers must ensure the category is not a
     * system category — system categories can only be archived
     * (см. § 25.1 Bible, open question № 5). The repository does not enforce
     * this rule; the use case does.
     */
    suspend fun softDelete(id: UUID)
}
